src/client.c

```diff
@@ -420,6 +420,41 @@ static bool finish_set_pool(PgSocket *client, bool takeover)
 	return ok;
 }
 
+bool check_user_connection_count(PgSocket *client)
+{
+	int client_connection_count;
+	int max_user_client_connections;
+
+	/* Check client_connection count limit */
+	if (!client->login_user_credentials)
+		return true;
+
+	if (!client->login_user_credentials->global_user)
+		return true;
+
+	if (!client->user_connection_counted) {
```

Comment:
I agree this is not pretty. But given the amount of counting bugs found in earlier versions, I prefer this way because this one is obviously correct and will stay correct in the future. I added a check for user_connection_counted in the decrement path too.